### ISSUE TITLE
libbindgen: unable to use constructors

### ISSUE THREAD

**Issue by bbigras** [servo/rust-bindgen#348]
I'm probably doing something wrong but it's weird that I can use `cv_String` and `new1()` is defined but not `cv::Mat::Mat(int, int, int)`. I tried to link against everything I got with `pkg-config --libs opencv`.

`constructor/src/lib.rs:6` is:
```rust
    include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
```

libbindgen 673cb5ff7555f2c026f4fbb6b8dfac0ff19e8725

wrapper.hpp
```c
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/imgcodecs/imgcodecs.hpp>
```

build.rs
```rust
extern crate libbindgen;
use libbindgen::Builder;

use std::env;
use std::path::PathBuf;

fn main() {
    println!("cargo:rustc-link-lib=opencv_stitching");
    println!("cargo:rustc-link-lib=opencv_superres");
    println!("cargo:rustc-link-lib=opencv_videostab");
    println!("cargo:rustc-link-lib=opencv_aruco");
    println!("cargo:rustc-link-lib=opencv_bgsegm");
    println!("cargo:rustc-link-lib=opencv_bioinspired");
    println!("cargo:rustc-link-lib=opencv_ccalib");
    println!("cargo:rustc-link-lib=opencv_dnn");
    println!("cargo:rustc-link-lib=opencv_dpm");
    println!("cargo:rustc-link-lib=opencv_fuzzy");
    println!("cargo:rustc-link-lib=opencv_hdf");
    println!("cargo:rustc-link-lib=opencv_line_descriptor");
    println!("cargo:rustc-link-lib=opencv_optflow");
    println!("cargo:rustc-link-lib=opencv_plot");
    println!("cargo:rustc-link-lib=opencv_reg");
    println!("cargo:rustc-link-lib=opencv_saliency");
    println!("cargo:rustc-link-lib=opencv_stereo");
    println!("cargo:rustc-link-lib=opencv_structured_light");
    println!("cargo:rustc-link-lib=opencv_rgbd");
    println!("cargo:rustc-link-lib=opencv_surface_matching");
    println!("cargo:rustc-link-lib=opencv_tracking");
    println!("cargo:rustc-link-lib=opencv_datasets");
    println!("cargo:rustc-link-lib=opencv_text");
    println!("cargo:rustc-link-lib=opencv_face");
    println!("cargo:rustc-link-lib=opencv_xfeatures2d");
    println!("cargo:rustc-link-lib=opencv_shape");
    println!("cargo:rustc-link-lib=opencv_video");
    println!("cargo:rustc-link-lib=opencv_ximgproc");
    println!("cargo:rustc-link-lib=opencv_calib3d");
    println!("cargo:rustc-link-lib=opencv_features2d");
    println!("cargo:rustc-link-lib=opencv_flann");
    println!("cargo:rustc-link-lib=opencv_xobjdetect");
    println!("cargo:rustc-link-lib=opencv_objdetect");
    println!("cargo:rustc-link-lib=opencv_ml");
    println!("cargo:rustc-link-lib=opencv_xphoto");
    println!("cargo:rustc-link-lib=opencv_highgui");
    println!("cargo:rustc-link-lib=opencv_videoio");
    println!("cargo:rustc-link-lib=opencv_imgcodecs");
    println!("cargo:rustc-link-lib=opencv_photo");
    println!("cargo:rustc-link-lib=opencv_imgproc");
    println!("cargo:rustc-link-lib=opencv_core");

    let bindings = Builder::default()
        .no_unstable_rust()
        .header("wrapper.hpp")       
        .opaque_type("std::char_traits")
        .no_unstable_rust()
        .whitelisted_function("cv.*")
        .whitelisted_type("Cv.*")
        .whitelisted_type("cv::.*")
        .generate()
        .expect("Unable to generate bindings");

    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
    bindings.write_to_file(out_path.join("bindings.rs"))
        .expect("Couldn't write bindings!");
}
```

lib.rs
```rust
#![allow(non_upper_case_globals)]
#![allow(non_camel_case_types)]
#![allow(non_snake_case)]

mod bindings {
    include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
}

#[test]
fn it_works() {
    use std::ffi::CString;

    let cstr = CString::new("frog.jpg").unwrap();

    let string = bindings::cv_String {
        cstr_: cstr.into_raw(),
        len_: 8,
    };

    unsafe {
        bindings::cv_imread(&string, 0);
    }

    let m1 = unsafe { bindings::cv_Mat::new1(1, 1, 1) };
}
```

output
```
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/home/bbigras/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/home/bbigras/dev/rust/tests/constructor/target/debug/deps/constructor-86500b1b61dc54aa.0.o" "-o" "/home/bbigras/dev/rust/tests/constructor/target/debug/deps/constructor-86500b1b61dc54aa" "-Wl,--gc-sections" "-pie" "-nodefaultlibs" "-L" "/home/bbigras/dev/rust/tests/constructor/target/debug/deps" "-L" "/home/bbigras/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "-Wl,-Bdynamic" "-l" "opencv_stitching" "-l" "opencv_superres" "-l" "opencv_videostab" "-l" "opencv_aruco" "-l" "opencv_bgsegm" "-l" "opencv_bioinspired" "-l" "opencv_ccalib" "-l" "opencv_dnn" "-l" "opencv_dpm" "-l" "opencv_fuzzy" "-l" "opencv_hdf" "-l" "opencv_line_descriptor" "-l" "opencv_optflow" "-l" "opencv_plot" "-l" "opencv_reg" "-l" "opencv_saliency" "-l" "opencv_stereo" "-l" "opencv_structured_light" "-l" "opencv_rgbd" "-l" "opencv_surface_matching" "-l" "opencv_tracking" "-l" "opencv_datasets" "-l" "opencv_text" "-l" "opencv_face" "-l" "opencv_xfeatures2d" "-l" "opencv_shape" "-l" "opencv_video" "-l" "opencv_ximgproc" "-l" "opencv_calib3d" "-l" "opencv_features2d" "-l" "opencv_flann" "-l" "opencv_xobjdetect" "-l" "opencv_objdetect" "-l" "opencv_ml" "-l" "opencv_xphoto" "-l" "opencv_highgui" "-l" "opencv_videoio" "-l" "opencv_imgcodecs" "-l" "opencv_photo" "-l" "opencv_imgproc" "-l" "opencv_core" "/home/bbigras/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libtest-1357b93f.rlib" "/home/bbigras/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgetopts-1357b93f.rlib" "/home/bbigras/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libterm-1357b93f.rlib" "/home/bbigras/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-1357b93f.rlib" "/home/bbigras/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-1357b93f.rlib" "/home/bbigras/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-1357b93f.rlib" "/home/bbigras/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librand-1357b93f.rlib" "/home/bbigras/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcollections-1357b93f.rlib" "/home/bbigras/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-1357b93f.rlib" "/home/bbigras/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc_jemalloc-1357b93f.rlib" "/home/bbigras/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-1357b93f.rlib" "/home/bbigras/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_unicode-1357b93f.rlib" "/home/bbigras/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-1357b93f.rlib" "/home/bbigras/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-1357b93f.rlib" "-l" "dl" "-l" "pthread" "-l" "gcc_s" "-l" "pthread" "-l" "c" "-l" "m" "-l" "rt" "-l" "util"
  = note: /home/bbigras/dev/rust/tests/constructor/target/debug/deps/constructor-86500b1b61dc54aa.0.o: In function `constructor::bindings::{{impl}}::new1':
/home/bbigras/dev/rust/tests/constructor/src/lib.rs:6: undefined reference to `cv::Mat::Mat(int, int, int)'
collect2: error: ld returned 1 exit status
```

[bindings.zip](https://github.com/servo/rust-bindgen/files/655943/bindings.zip)

**Comment by emilio** [comment]
That's fun.

So that constructor is defined as:

```cpp
inline
Mat::Mat(int _rows, int _cols, int _type)
    : flags(MAGIC_VAL), dims(0), rows(0), cols(0), data(0), datastart(0), dataend(0),
      datalimit(0), allocator(0), u(0), size(&rows)
{
    create(_rows, _cols, _type);
}
```

Which means that it's an inline function, so it wouldn't generate any symbol rust can call into. We're supposed to detect that, but apparently we don't when it's in another file?

I think that can legitimately be considered a libclang bug. We can work around it (though it's kind of an expensive check, plus annoying)...

TL;DR: That constructor shouldn't have been generated in the first place.

**Comment by kkimdev** [comment]
This seems like it can be a painful restriction in some cases.  Ideally, maybe we need a pre-step that compiles inline functions concretely, and exposing non-lined API.

**Comment by emilio** [comment]
That's fair. I think you can run bindgen and compile with `-fno-inline-functions` (passing that flag also to bindgen) and it should just work.

I wonder if there's an easy way of compiling a staticlib inlining functions in C++ code, but force the compiler to spit symbols as `-fno-inline-functions` would do. That'd be a nice solution for this problem I think.

**Comment by fitzgen** [comment]
I think we can close this issue now? It doesn't look like we have anything actionable here. If I'm missing something, please re-open!
